File: src/decoders/notebook-punishment.ts
import type { NotebookPunishment, SessionHandle } from "~/models";
import { decodePronoteDate } from "./pronote-date";
import { decodeAttachment } from "./attachment";

export const decodeNotebookPunishment = (punishment: any, session: SessionHandle): NotebookPunishment => {
  return {
    id: punishment.N,
    title: punishment.nature.V.L,
    reasons: punishment.listeMotifs.V.map((motif: any) => motif.L),

    exclusion: punishment.estUneExclusion,
    isDuringLesson: !punishment.horsCours,

    workToDo: punishment.travailAFaire,
    workToDoDocuments: punishment.documentsTAF.V.map((document: any) => decodeAttachment(document, session)),

    circumstances: punishment.circonstances,
    circumstancesDocuments: punishment.documentsCirconstances.V.map((document: any) => decodeAttachment(document, session)),

    giver: punishment.demandeur.V.L,
    dateGiven: decodePronoteDate(punishment.dateDemande.V),

    durationMinutes: punishment.duree
  };
};

File: library/src/commonMain/kotlin/decoders/notebook-punishment.kt
package decoders

import kotlinx.serialization.json.*
import models.NotebookPunishment
import models.SessionHandle

fun decodeNotebookPunishment (punishment: JsonObject, session: SessionHandle) = NotebookPunishment(
    id = punishment["N"]!!.jsonPrimitive.content,
    title = punishment["nature"]!!.jsonObject["V"]!!.jsonObject["L"]!!.jsonPrimitive.content,
    reasons = punishment["listeMotifs"]!!.jsonObject["V"]!!.jsonArray.map { it.jsonObject["L"]!!.jsonPrimitive.content },

    exclusion = punishment["estUneExclusion"]!!.jsonPrimitive.boolean,
    isDuringLesson = !punishment["horsCours"]!!.jsonPrimitive.boolean,

    workToDo = punishment["travailAFaire"]!!.jsonPrimitive.content,
    workToDoDocuments = punishment["documentsTAF"]!!.jsonObject["V"]!!.jsonArray.map { decodeAttachment(it.jsonObject, session.information) },

    circumstances = punishment["circonstances"]!!.jsonPrimitive.content,
    circumstancesDocuments = punishment["documentsCirconstances"]!!.jsonObject["V"]!!.jsonArray.map { decodeAttachment(it.jsonObject, session.information) },

    giver = punishment["demandeur"]!!.jsonObject["V"]!!.jsonObject["L"]!!.jsonPrimitive.content,
    dateGiven = decodePronoteDate(punishment["dateDemande"]!!.jsonObject["V"]!!.jsonPrimitive.content),

    durationMinutes = punishment["duree"]!!.jsonPrimitive.int
)